File: frontend/src/lib/api/groups.ts
import { api } from './client';
import type { Paper } from './papers';
import { papersApi } from './papers';

export type { Paper } from './papers';

export interface Group {
  id: number;
  name: string;
  parent_id?: number;
  created_at: string;
  updated_at: string;
  papers?: Paper[];
  children?: Group[];
}

export interface GroupCreate {
  name: string;
  parent_id?: number;
}

export interface GroupUpdate {
  name?: string;
  parent_id?: number;
}

export const groupsApi = {
  list: async (): Promise<Group[]> => {
    return api.get<Group[]>('/groups');
  },

  get: async (id: number): Promise<Group> => {
    return api.get<Group>(`/groups/${id}`);
  },

  create: async (group: GroupCreate): Promise<Group> => {
    return api.post<Group>('/groups', group);
  },

  update: async (id: number, updates: GroupUpdate): Promise<Group> => {
    return api.patch<Group>(`/groups/${id}`, updates);
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/groups/${id}`);
  },

  updatePaperGroups: async (paperId: number, groupIds: number[]): Promise<void> => {
    await papersApi.update(paperId, { group_ids: groupIds });
  },
};

File: backend/app/api/groups.py
"""Groups API endpoints."""

from typing import List

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.crud import (
  create_group,
  delete_group,
  get_group_or_404,
  list_groups,
  update_group,
)
from app.dependencies import get_db
from app.schemas.group import Group as GroupSchema
from app.schemas.group import GroupCreate, GroupUpdate

router = APIRouter()


@router.get("/groups", response_model=List[GroupSchema])
async def list_groups_endpoint(session: AsyncSession = Depends(get_db)):
  """List all groups with their papers and children."""
  groups = await list_groups(session)
  return [GroupSchema.model_validate(g) for g in groups]


@router.post("/groups", response_model=GroupSchema, status_code=201)
async def create_group_endpoint(
  group_in: GroupCreate, session: AsyncSession = Depends(get_db)
):
  """Create a new group."""
  group = await create_group(session, group_in.name, group_in.parent_id)
  return GroupSchema.model_validate(group)


@router.get("/groups/{group_id}", response_model=GroupSchema)
async def get_group_endpoint(group_id: int, session: AsyncSession = Depends(get_db)):
  """Get a single group by ID."""
  group = await get_group_or_404(session, group_id, with_relations=True)
  return GroupSchema.model_validate(group)


@router.patch("/groups/{group_id}", response_model=GroupSchema)
async def update_group_endpoint(
  group_id: int, group_update: GroupUpdate, session: AsyncSession = Depends(get_db)
):
  """Update a group."""
  group = await update_group(
    session,
    group_id,
    name=group_update.name,
    parent_id=group_update.parent_id,
  )
  return GroupSchema.model_validate(group)


@router.delete("/groups/{group_id}", status_code=204)
async def delete_group_endpoint(group_id: int, session: AsyncSession = Depends(get_db)):
  """Delete a group."""
  await delete_group(session, group_id)
  return None
